ccb/upstream_project.py

```diff
@@ -262,10 +311,49 @@ def __init__(self, recipe):
         self.domain = domain
         self.owner = owner
         self.repo = repo
+        self.__versions = None
+
+    async def versions(self):
+        if self.__versions is None:
+            try:
+                async with aiohttp.ClientSession(raise_for_status=True) as client:
+                    async with client.get(
+                        f"https://{self.domain}/api/v4/projects/{self.owner}%2F{self.repo}/releases"
+                    ) as resp:
+                        def _get_url_for_version(v):
+                            artifacts = v["assets"]["links"]
+                            for a in artifacts:
+                                if a["direct_asset_url"].endswith(".tar.xz"):
+                                    return a["direct_asset_url"]
+                            for a in artifacts:
+                                if a["direct_asset_url"].endswith(".tar.gz"):
+                                    return a["direct_asset_url"]
+                            for a in artifacts:
+                                if a["direct_asset_url"].endswith(".tar.bz2"):
+                                    return a["direct_asset_url"]
+                            return f"https://{self.domain}/{self.owner}/{self.repo}/-/archive/{v['tag_name']}/{self.repo}-{v['tag_name']}.tar.gz"
```

Comment:
If I read this right, this is a default url, does it always work ?
And if it does, why not using it all the time ?
Same question for GitHub ?

Reply:
If I'm not mistaken in the tag archive, which is used as a fallback if we did not find a usable release tarball.